docs-user/bunny-2.md

```diff
@@ -0,0 +1,78 @@
+      // Draw the current set of rectangles.
+      drawRects(message.data.drawCalls);
+      // Post a message to the worker asking for the next draw calls.
+      worker.postMessage({ type: 'generate-draw-calls' });
+    });
+  }
+});
+```
+
+In this example, the iframe draws to the canvas, and once it's done, it posts a message to the worker asking it generate a new list of the draw calls. While this works, it's not harnessing the fact that the worker can run code in parallel. The fix luckily is simple. Swap the order of the `drawRect` and `worker.postMessage` calls.
+
+```js
+worker.addEventListener('message', message => {
+  if (message.data.type === 'draw') {
+    requestAnimationFrame(() => {
+      // Ask the worker to generate the new draw calls for the NEXT frame.
+      worker.postMessage({ type: 'generate-draw-calls' });
+      // Draw the current frame.
+      drawRects(message.data.drawCalls);
+    });
+  }
+});
+```
+
+## The new profile
+
+Profiling this change reveals the new behavior: https://perfht.ml/2IpSRr4
+
+![A view of the content and worker stacks. The content thread and worker thread do work in parallel, and are not blocked by each other.](./images/bunny-analysis/threads-parallel.png)
+
+Now the content process' main thread is not blocked by the worker. The overall time to draw a single frame is the same, but conceptually this has an important difference. The blank spots in the content thread no longer represent time where it is blocked. Now it truly represents idle time, where no work needs to be done.
+
+## How this improves performance
+
+Consider what happens if the worker takes a long time to compute the new draw calls. As the work scales, it happens in parallel. To simulate this, a function called `doWork` can be added to the worker.
+
+## Busy synchronous threads
+
+This is a profile of what this looks like for a synchronous ordering: https://perfht.ml/2wuFSj1
+
+![A view of the content and worker stacks. The content thread and worker thread do work one after the other. There is a large gap between frames.](./images/bunny-analysis/threads-work-sync.png)
+
+The worker takes long enough that the frames start to skip between each rendering. The result is that both threads and up with lots of idle time, and the end-user perceives that the performance is slower.
+
+## Busy parallel threads
+
+This is a profile for parallel processing: https://perfht.ml/2KajgWV
+
+![A view of the content and worker stacks. The content thread and worker thread do work in parallel, and are not blocked by each other.](./images/bunny-analysis/threads-work-parallel.png)
+
+Here there are a lot more frames being drawn. The end-user will perceive a more fluid animation, and the experience will feel much faster. This performance boost perceived by the end user means that in fact, the threads are each doing more work. The work is evenly distributed between the two threads, and they do spend time idle by being blocked.
```

Comment:
Would there be some advantage in starting more Workers if the work is even longer ? Hypothetical question.

Reply:
Not in this case.